tests/test_crawlers.py

```diff
@@ -561,90 +561,98 @@ def test_get_download_url_no_direct_download_link(self, mock_get_link, mock_http
-        self.mock_requests_get.side_effect = self.requests_get_side_effect
+        with mock.patch.object(self.crawler, '_get_next_page', return_value=''):
+            self.assertListEqual(list(self.crawler), expected_urls)
+
+    def test_get_next_page(self):
+        """_get_next_page() should get the page at the current offset,
+        then increment the offset
+        """
+        with mock.patch.object(self.crawler, '_http_get', return_value='foo'):
+            with self.assertLogs(self.crawler.LOGGER, level=logging.INFO):
+                self.assertEqual(self.crawler._get_next_page(), 'foo')
+                self.assertEqual(self.crawler.request_parameters['params']['offset'], 1)
+
+    def test_abstract_get_datasets_info(self):
+        """_get_datasets_info() should raise a NotImplementedError
+        when called directly from HTTPPaginatedAPICrawler
+        """
+        crawler = crawlers.HTTPPaginatedAPICrawler('foo')
+        with self.assertRaises(NotImplementedError):
+            crawler._get_datasets_info('')
 
-        # Initialize a list of opened files which will be closed in tearDown()
-        self.opened_files = []
 
+class CopernicusOpenSearchAPICrawlerTestCase(unittest.TestCase):
+    """Tests for the Copernicus OpenSearch API crawler"""
+    BASE_URL = 'https://scihub.copernicus.eu/dhus/search'
+    SEARCH_TERMS = '(platformname:Sentinel-1 OR platformname:Sentinel-3) AND NOT L0'
+    PAGE_SIZE = 2
+
+    def setUp(self):
         self.crawler = crawlers.CopernicusOpenSearchAPICrawler(
             url=self.BASE_URL, search_terms=self.SEARCH_TERMS, username='user', password='pass',
             page_size=self.PAGE_SIZE, initial_offset=0)
 
-    def tearDown(self):
-        self.patcher_requests_get.stop()
-        # Close any files opened during the test
-        for opened_file in self.opened_files:
-            opened_file.close()
-
-    def test_instantiation(self):
-        """Test the correct instantiation of a Copernicus OpenSearch API crawler"""
-        self.assertIsInstance(self.crawler, crawlers.Crawler)
-        self.assertEqual(self.crawler.url, self.BASE_URL)
-        self.assertEqual(self.crawler.initial_offset, 0)
-        self.assertDictEqual(self.crawler.request_parameters, {
-            'params': {
-                'q': f"({self.SEARCH_TERMS}) AND (beginposition:[1-01-01T00:00:00Z TO NOW])",
-                'start': 0,
-                'rows': self.PAGE_SIZE,
-                'orderby': 'beginposition asc'
-            },
-            'auth': ('user', 'pass')
-        })
-        self.assertEqual(self.crawler._urls, [])
-
     def test_build_parameters_with_standard_time_range(self):
         """Build the request parameters with a time range composed of two datetime objects"""
-        request_parameters = crawlers.CopernicusOpenSearchAPICrawler._build_request_parameters(
+        request_parameters = self.crawler._build_request_parameters(
```

Comment:
You define a class inside another class in a nested manner. Please do it in the setup method of tests in order to prevent tests and actions inside each test manipulating other tests and their actions.

Reply:
There is no nested class in the `CopernicusOpenSearchAPICrawlerTestCase`, and the `self.crawler` attribute used on this line is defined in the `setUp()` method.